components/Discussion/NotificationOptions.js

```diff
@@ -0,0 +1,253 @@
+  })
+}
+
+class NotificationOptions extends PureComponent {
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      mutating: false,
+      webNotificationsPermission: null
+    }
+  }
+
+  componentDidMount () {
+    this.subscribe()
+    this.maybeMute()
+  }
+
+  componentDidUpdate () {
+    this.subscribe()
+    this.maybeMute()
+  }
+
+  initNotificationsState () {
+    this.setState({
+      webNotificationsPermission: getNotificationPermission()
+    })
+  }
+
+  confirmPermission () {
+    if (!isNotificationSupported()) {
+      return
+    }
+    const { t, data, updateNotificationSettings } = this.props
+    const { discussionNotificationChannels } = data && data.me
+
+    window.Notification.requestPermission((status) => {
+      if (status !== 'granted') {
+        return
+      }
+      // Globally opt in to the WEB notification channel.
+      const channels = discussionNotificationChannels.indexOf('WEB') === -1
+        ? [...discussionNotificationChannels, 'WEB']
+        : discussionNotificationChannels
+      updateNotificationSettings({
+        discussionNotificationChannels: channels
+      })
+      /* eslint-disable no-new */
+      new window.Notification(t('components/Discussion/WelcomeNotification/title'), {
+        body: t('components/Discussion/WelcomeNotification/body'),
+        icon: `${CDN_FRONTEND_BASE_URL}/frontend/static/apple-touch-icon.png`
+      })
+    })
+  }
+
+  subscribe () {
+    this.initNotificationsState()
+    if (this.unsubscribe || !isNotificationSupported() || !this.props.data) {
+      return
+    }
+    this.unsubscribe = this.props.data.subscribeToMore({
+      document: webNotificationSubscription,
+      updateQuery: (prev, { subscriptionData }) => {
+        const webNotification = subscriptionData.data && subscriptionData.data.webNotification
+        if (!webNotification || this.state.webNotificationsPermission === 'denied') {
+          return prev
+        }
+        /* eslint-disable no-new */
+        const n = new window.Notification(webNotification.title, {
+          body: webNotification.body,
+          icon: webNotification.icon
+        })
+        n.onclick = (e) => {
+          e.preventDefault() // prevent the browser from focusing the Notification's tab
+          window.focus()
+          window.location = webNotification.url
+        }
+        return prev
+      }
+    })
+  }
+
+  maybeMute () {
+    const {
+      data: { discussion },
+      setDiscussionPreferences,
+      mute
+    } = this.props
+    if (!mute || !discussion || this.state.mutating) {
+      return
+    }
+
+    const {
+      userPreference
+    } = discussion
+
+    // Mute notifications for this discussion if not already done
+    if (!userPreference || userPreference.notifications !== 'NONE') {
+      this.setState({
+        mutating: true
+      })
+
+      // Preserve existing user preferences.
+      const anonymity = userPreference ? userPreference.anonymity : false
```

Comment:
this is a wrong assumption—a discussion can enforce that user comment with anonymity true. This then fails with «In dieser Debatte sind nur anonyme Kommentare zugelassen.»–also we always should handle errors!